File: apps/web/src/api/analytics.ts
export interface AnalyticsFilters {
  projectId: string;
  dateFrom?: string;
  dateTo?: string;
  sourceSystem?: string;
  intakeChannelCode?: string;
  affectedChannelCode?: string;
  locationId?: string;
  locationScope?: string;
  customerLifecycleStageCode?: string;
  customerLifecycleStepCode?: string;
  touchpointCode?: string;
  serviceRequestStepCode?: string;
  serviceCode?: string;
  issueCode?: string;
  sentiment?: string;
  operationalSeverity?: string;
}

export interface AnalyticsSummary {
  itemVolume: number;
  csatScore: number;
  positiveRate: number;
  negativeRate: number;
  unknownRate: number;
  activeHotspots: number;
}

export interface AnalyticsTrendPoint {
  bucket: string;
  itemVolume: number;
  negativeRate: number;
  unknownRate: number;
  activeHotspots: number;
}

export interface AnalyticsBreakdownItem {
  code: string;
  name: string;
  itemVolume: number;
  percentage: number;
  negativeRate: number;
  activeHotspots: number;
}

export interface AnalyticsFilterOption {
  code: string;
  name: string;
  id?: string;
}

export interface AnalyticsFilterOptions {
  sourceSystems: AnalyticsFilterOption[];
  intakeChannels: AnalyticsFilterOption[];
  affectedChannels: AnalyticsFilterOption[];
  locations: AnalyticsFilterOption[];
  journeyStages: AnalyticsFilterOption[];
  journeySteps: AnalyticsFilterOption[];
  touchpoints: AnalyticsFilterOption[];
  serviceRequestSteps: AnalyticsFilterOption[];
  services: AnalyticsFilterOption[];
  issues: AnalyticsFilterOption[];
  sentiments: AnalyticsFilterOption[];
  severities: AnalyticsFilterOption[];
}

interface ApiEnvelope<T> {
  data: T;
}

interface ApiSummary {
  item_volume: number;
  csat_score: number;
  positive_rate: number;
  negative_rate: number;
  unknown_rate: number;
  active_hotspots: number;
}

interface ApiTrendPoint {
  bucket: string;
  item_volume: number;
  negative_rate: number;
  unknown_rate: number;
  active_hotspots: number;
}

interface ApiBreakdownItem {
  dimension: { code: string; name_vi: string };
  metrics: { item_volume: number; percentage: number; negative_rate: number; active_hotspots: number };
}

interface ApiFilterOption {
  code: string;
  name_vi: string;
  id: string | null;
}

interface ApiFilterOptions {
  source_systems: ApiFilterOption[];
  intake_channels: ApiFilterOption[];
  affected_channels: ApiFilterOption[];
  locations: ApiFilterOption[];
  journey_stages: ApiFilterOption[];
  journey_steps: ApiFilterOption[];
  touchpoints?: ApiFilterOption[];
  service_request_steps: ApiFilterOption[];
  services: ApiFilterOption[];
  issues: ApiFilterOption[];
  sentiments: ApiFilterOption[];
  severities: ApiFilterOption[];
}

export class AnalyticsApiError extends Error {
  constructor(message: string, public readonly status?: number) {
    super(message);
    this.name = 'AnalyticsApiError';
  }
}

const baseUrl = (import.meta.env.VITE_API_BASE_URL ?? 'http://localhost:8000').replace(/\/$/, '');
const projectId = import.meta.env.VITE_ANALYTICS_PROJECT_ID?.trim();

type ErrorDetail = {
  loc?: unknown;
  msg?: unknown;
};

const isErrorDetail = (value: unknown): value is ErrorDetail => (
  typeof value === 'object' && value !== null
);

/** Turn FastAPI's string or structured validation detail into readable UI text. */
export const formatAnalyticsApiError = (detail: unknown): string | null => {
  if (typeof detail === 'string' && detail.trim()) return detail;

  if (Array.isArray(detail)) {
    const messages = detail.flatMap((item) => {
      if (!isErrorDetail(item) || typeof item.msg !== 'string') return [];
      const location = Array.isArray(item.loc)
        ? item.loc.filter((part) => part !== 'query' && part !== 'body').join('.')
        : '';
      return [location ? `${location}: ${item.msg}` : item.msg];
    });
    return messages.length ? messages.join(' • ') : null;
  }

  if (isErrorDetail(detail) && typeof detail.msg === 'string') return detail.msg;
  return null;
};

/** Format a calendar date without converting it to UTC first. */
export const formatLocalDate = (value: Date): string => {
  const year = value.getFullYear();
  const month = String(value.getMonth() + 1).padStart(2, '0');
  const day = String(value.getDate()).padStart(2, '0');
  return `${year}-${month}-${day}`;
};

export const analyticsConfigurationError = projectId
  ? null
  : 'Thiếu VITE_ANALYTICS_PROJECT_ID. Hãy cấu hình UUID project trong apps/web/.env.local.';

export const defaultAnalyticsFilters = (): AnalyticsFilters | null => {
  if (!projectId) return null;
  const dateTo = new Date();
  const dateFrom = new Date();
  dateFrom.setDate(dateTo.getDate() - 29);
  return {
    projectId,
    dateFrom: formatLocalDate(dateFrom),
    dateTo: formatLocalDate(dateTo),
  };
};

const filterParams = (filters: AnalyticsFilters): URLSearchParams => {
  const params = new URLSearchParams({ project_id: filters.projectId });
  const names: Record<Exclude<keyof AnalyticsFilters, 'projectId'>, string> = {
    dateFrom: 'date_from',
    dateTo: 'date_to',
    sourceSystem: 'source_system',
    intakeChannelCode: 'intake_channel_code',
    affectedChannelCode: 'affected_channel_code',
    locationId: 'location_id',
    locationScope: 'location_scope',
    customerLifecycleStageCode: 'customer_lifecycle_stage_code',
    customerLifecycleStepCode: 'customer_lifecycle_step_code',
    touchpointCode: 'touchpoint_code',
    serviceRequestStepCode: 'service_request_step_code',
    serviceCode: 'service_code',
    issueCode: 'issue_code',
    sentiment: 'sentiment',
    operationalSeverity: 'operational_severity',
  };
  for (const [key, apiName] of Object.entries(names)) {
    const value = filters[key as keyof typeof names];
    if (value) params.set(apiName, value);
  }
  return params;
};

const request = async <T>(path: string, params: URLSearchParams): Promise<T> => {
  const response = await fetch(`${baseUrl}/api/v1/analytics${path}?${params.toString()}`);
  if (!response.ok) {
    let detail = `Analytics API trả về lỗi ${response.status}.`;
    try {
      const body = await response.json() as { detail?: unknown };
      detail = formatAnalyticsApiError(body.detail) ?? detail;
    } catch {
      // Keep the HTTP status message if the response body is not JSON.
    }
    throw new AnalyticsApiError(detail, response.status);
  }
  return response.json() as Promise<T>;
};

export const getAnalyticsSummary = async (filters: AnalyticsFilters): Promise<AnalyticsSummary> => {
  const response = await request<ApiEnvelope<ApiSummary>>('/summary', filterParams(filters));
  return {
    itemVolume: response.data.item_volume,
    csatScore: response.data.csat_score,
    positiveRate: response.data.positive_rate,
    negativeRate: response.data.negative_rate,
    unknownRate: response.data.unknown_rate,
    activeHotspots: response.data.active_hotspots,
  };
};

export const getAnalyticsTrend = async (
  filters: AnalyticsFilters,
  grain: 'day' | 'week' | 'month' = 'day',
): Promise<AnalyticsTrendPoint[]> => {
  const params = filterParams(filters);
  params.set('grain', grain);
  const response = await request<ApiEnvelope<ApiTrendPoint[]>>('/trend', params);
  return response.data.map((point) => ({
    bucket: point.bucket,
    itemVolume: point.item_volume,
    negativeRate: point.negative_rate,
    unknownRate: point.unknown_rate,
    activeHotspots: point.active_hotspots,
  }));
};

export const getAnalyticsBreakdown = async (
  filters: AnalyticsFilters,
  dimension: string,
  limit = 20,
): Promise<AnalyticsBreakdownItem[]> => {
  const params = filterParams(filters);
  params.set('dimension', dimension);
  params.set('limit', String(limit));
  const response = await request<ApiEnvelope<ApiBreakdownItem[]>>('/breakdown', params);
  return response.data.map((item) => ({
    code: item.dimension.code,
    name: item.dimension.name_vi,
    itemVolume: item.metrics.item_volume,
    percentage: item.metrics.percentage,
    negativeRate: item.metrics.negative_rate,
    activeHotspots: item.metrics.active_hotspots,
  }));
};

const toFilterOption = (item: ApiFilterOption): AnalyticsFilterOption => ({
  code: item.code,
  name: item.name_vi,
  ...(item.id ? { id: item.id } : {}),
});

export const getAnalyticsFilterOptions = async (
  filters: AnalyticsFilters,
): Promise<AnalyticsFilterOptions> => {
  const response = await request<ApiEnvelope<ApiFilterOptions>>('/filter-options', filterParams(filters));
  const data = response.data;
  return {
    sourceSystems: (data.source_systems ?? []).map(toFilterOption),
    intakeChannels: (data.intake_channels ?? []).map(toFilterOption),
    affectedChannels: (data.affected_channels ?? []).map(toFilterOption),
    locations: (data.locations ?? []).map(toFilterOption),
    journeyStages: (data.journey_stages ?? []).map(toFilterOption),
    journeySteps: (data.journey_steps ?? []).map(toFilterOption),
    touchpoints: (data.touchpoints ?? []).map(toFilterOption),
    serviceRequestSteps: (data.service_request_steps ?? []).map(toFilterOption),
    services: (data.services ?? []).map(toFilterOption),
    issues: (data.issues ?? []).map(toFilterOption),
    sentiments: (data.sentiments ?? []).map(toFilterOption),
    severities: (data.severities ?? []).map(toFilterOption),
  };
};

File: apps/api/routers/analytics.py
"""P0 analytics HTTP endpoints backed by the governed semantic view."""
from __future__ import annotations

from datetime import date
from typing import Annotated, Literal
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status

from apps.api.deps import get_analytics_repository
from apps.api.schemas.analytics import (
    AnalyticsMeta,
    BreakdownDimension,
    BreakdownItem,
    BreakdownMetrics,
    BreakdownResponse,
    DataQualityResponse,
    AnalyticsFilterOptions,
    SummaryData,
    SummaryResponse,
    TrendPoint,
    TrendResponse,
    FilterOption,
    FilterOptionsResponse,
)
from packages.application.analytics.analytics_service import AnalyticsFilters
from packages.domain.shared.enums import OperationalSeverity, Sentiment
from packages.infrastructure.db.repositories.analytics import AnalyticsRepository

router = APIRouter(prefix="/api/v1/analytics", tags=["Analytics"])

AnalyticsRepositoryDep = Annotated[AnalyticsRepository, Depends(get_analytics_repository)]
AnalyticsGrain = Literal["day", "week", "month"]
AnalyticsDimension = Literal[
    "service",
    "issue",
    "location",
    "journey_stage",
    "journey_step",
    "touchpoint",
    "service_request_step",
    "intake_channel",
    "affected_channel",
    "sentiment",
    "severity",
]


def shared_filters(
    project_id: UUID,
    date_from: date | None = None,
    date_to: date | None = None,
    source_system: str | None = None,
    intake_channel_code: str | None = None,
    affected_channel_code: str | None = None,
    location_id: UUID | None = None,
    location_scope: UUID | None = None,
    customer_lifecycle_stage_code: str | None = None,
    customer_lifecycle_step_code: str | None = None,
    touchpoint_code: str | None = None,
    service_request_step_code: str | None = None,
    service_code: str | None = None,
    issue_code: str | None = None,
    sentiment: Sentiment | None = None,
    operational_severity: OperationalSeverity | None = None,
) -> AnalyticsFilters:
    """Parse the one shared P0 dashboard filter context for every route."""
    if date_from is not None and date_to is not None and date_from > date_to:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_CONTENT,
            detail="date_from must not be later than date_to",
        )
    return AnalyticsFilters(
        project_id=project_id,
        date_from=date_from,
        date_to=date_to,
        source_system=source_system,
        intake_channel_code=intake_channel_code,
        affected_channel_code=affected_channel_code,
        location_id=location_id,
        location_scope=location_scope,
        customer_lifecycle_stage_code=customer_lifecycle_stage_code,
        customer_lifecycle_step_code=customer_lifecycle_step_code,
        touchpoint_code=touchpoint_code,
        service_request_step_code=service_request_step_code,
        service_code=service_code,
        issue_code=issue_code,
        sentiment=sentiment,
        operational_severity=operational_severity,
    )


AnalyticsFiltersDep = Annotated[AnalyticsFilters, Depends(shared_filters)]


def _meta(filters: AnalyticsFilters) -> AnalyticsMeta:
    """Return JSON-safe filter metadata for a future drill-down consumer."""
    return AnalyticsMeta(
        filter_context={
            name: str(value) if value is not None else None
            for name, value in (
                ("project_id", filters.project_id),
                ("date_from", filters.date_from),
                ("date_to", filters.date_to),
                ("source_system", filters.source_system),
                ("intake_channel_code", filters.intake_channel_code),
                ("affected_channel_code", filters.affected_channel_code),
                ("location_id", filters.location_id),
                ("location_scope", filters.location_scope),
                ("customer_lifecycle_stage_code", filters.customer_lifecycle_stage_code),
                ("customer_lifecycle_step_code", filters.customer_lifecycle_step_code),
                ("touchpoint_code", filters.touchpoint_code),
                ("service_request_step_code", filters.service_request_step_code),
                ("service_code", filters.service_code),
                ("issue_code", filters.issue_code),
                ("sentiment", filters.sentiment),
                ("operational_severity", filters.operational_severity),
            )
        },
    )


@router.get(
    "/summary",
    response_model=SummaryResponse,
    operation_id="getAnalyticsSummary",
)
async def get_summary(
    filters: AnalyticsFiltersDep,
    repository: AnalyticsRepositoryDep,
) -> SummaryResponse:
    """Return eligible feedback-item volume and sentiment-derived rates."""
    summary = await repository.get_summary(filters)
    return SummaryResponse(
        data=SummaryData(
            item_volume=summary.total,
            csat_score=summary.csat_score,
            positive_rate=summary.positive_rate,
            negative_rate=summary.negative_rate,
            unknown_rate=summary.unknown_rate,
            active_hotspots=summary.active_hotspots,
        ),
        meta=_meta(filters),
    )


@router.get("/trend", response_model=TrendResponse, operation_id="getAnalyticsTrend")
async def get_trend(
    filters: AnalyticsFiltersDep,
    repository: AnalyticsRepositoryDep,
    grain: AnalyticsGrain = Query("day"),
    metric: Literal["item_volume", "negative_rate", "unknown_rate", "active_hotspots"] = Query("item_volume"),
) -> TrendResponse:
    """Return P0 item-volume trend by day, week, or month."""
    del metric  # All P0 trend metrics are returned from the same governed bucket series.
    trend = await repository.get_trend(filters, grain)
    return TrendResponse(
        data=[
            TrendPoint(
                bucket=point.time_bucket,
                item_volume=point.volume,
                negative_rate=point.negative_rate,
                unknown_rate=point.unknown_rate,
                active_hotspots=point.active_hotspots,
            )
            for point in trend
        ],
        meta=_meta(filters),
    )


@router.get(
    "/breakdown",
    response_model=BreakdownResponse,
    operation_id="getAnalyticsBreakdown",
)
async def get_breakdown(
    filters: AnalyticsFiltersDep,
    repository: AnalyticsRepositoryDep,
    dimension: AnalyticsDimension,
    limit: Annotated[int, Query(ge=1, le=100)] = 20,
) -> BreakdownResponse:
    """Return a dimension breakdown from the same shared filter context."""
    breakdown = await repository.get_breakdown(filters, dimension)
    return BreakdownResponse(
        data=[
            BreakdownItem(
                dimension=BreakdownDimension(
                    code=item.dimension_key,
                    name_vi=item.dimension_name or item.dimension_key,
                ),
                metrics=BreakdownMetrics(
                    item_volume=item.count,
                    percentage=item.percentage,
                    negative_rate=item.negative_rate,
                    active_hotspots=item.active_hotspots,
                ),
            )
            for item in breakdown[:limit]
        ],
        meta=_meta(filters),
    )


@router.get(
    "/filter-options",
    response_model=FilterOptionsResponse,
    operation_id="getAnalyticsFilterOptions",
)
async def get_filter_options(
    filters: AnalyticsFiltersDep,
    repository: AnalyticsRepositoryDep,
) -> FilterOptionsResponse:
    """Return human-readable, data-backed values for the shared P0 filters."""
    options = await repository.get_filter_options(filters)
    return FilterOptionsResponse(
        data=AnalyticsFilterOptions(
            **{
                field: [FilterOption(code=item.code, name_vi=item.name, id=item.id) for item in values]
                for field, values in options.items()
            }
        ),
        meta=_meta(filters),
    )


@router.get(
    "/data-quality",
    response_model=DataQualityResponse,
    status_code=status.HTTP_501_NOT_IMPLEMENTED,
    operation_id="getAnalyticsDataQuality",
)
async def get_data_quality(filters: AnalyticsFiltersDep) -> DataQualityResponse:
    """Reserve the route until a separate non-eligible data quality mart exists."""
    del filters
    raise HTTPException(
        status_code=status.HTTP_501_NOT_IMPLEMENTED,
        detail=(
            "data-quality requires a dedicated quality mart because the governed "
            "analytics view contains eligible items only"
        ),
    )
